### PR TITLE
Use string for metadata

### PR DESCRIPTION
**Pull request by pscott**
This PR adds support for the metadata hash (defined as a string instead of felt).

### CODE REVIEW

File: test/starknet/vanilla_space.ts [hunk 1]
```diff
@@ -48,31 +49,59 @@ async function setup() {
...
         to: space_contract,
         function_selector: BigInt(getSelectorFromName(PROPOSAL_METHOD)),
-        calldata: [
-          proposer_address,
-          execution_hash,
-          metadata_uri,
-          eth_block_number,
-          BigInt(params.length),
-        ],
+        calldata: calldata,
```
Comment (bonustrack) [inline test/starknet/vanilla_space.ts]:
```suggestion
        calldata,
```

File: test/starknet/vanilla_space.ts [hunk 2]
```diff
@@ -48,31 +49,59 @@ async function setup() {
   };
 }
 
+function setup_calldata(
```
Comment (Orland0x) [inline test/starknet/vanilla_space.ts]: camelCase here?

File: test/starknet/vanilla_space.ts [hunk 3]
```diff
@@ -48,31 +49,59 @@ async function setup() {
...
+  }
+
+  calldata.push(eth_block_number);
+  calldata.push(BigInt(params.length));
+
+  for (let i = 0; i < params.length; i++) {
+    calldata.push(params[i]);
+  }
+
+  return calldata;
```
Comment (bonustrack) [inline test/starknet/vanilla_space.ts]: Try this, that's much more readable
```suggestion
  return [
    proposer_address,
    execution_hash,
    BigInt(metadata_uri.length),
    ...metadata_uri,
    eth_block_number,
    BigInt(params.length),
    ...params
  ];
```

File: test/starknet/vanilla_space.ts (hunk 2 above)
Comment (pscott) [reply test/starknet/vanilla_space.ts]: [use camelcase; rename setup_calldata to setupProposeCalldata](https://github.com/snapshot-labs/sx-core/pull/54/commits/9a1ad38aecf21c32382abf5a53c2387beff40eb2)

File: test/starknet/vanilla_space.ts (hunk 1 above)
Comment (pscott) [reply test/starknet/vanilla_space.ts]: [use camelcase; rename setup_calldata to setupProposeCalldata](https://github.com/snapshot-labs/sx-core/pull/54/commits/9a1ad38aecf21c32382abf5a53c2387beff40eb2)

File: test/starknet/vanilla_space.ts (hunk 2 above)
Comment (bonustrack) [reply test/starknet/vanilla_space.ts]: Yep, we should use camelCase on all the functions name but also all variables (for TS)

File: test/starknet/vanilla_space.ts (hunk 3 above)
Comment (bonustrack) [reply test/starknet/vanilla_space.ts]: (You may not need a function in the end)

**Comment by Orland0x** [comment]
looks good

**Review by Orland0x — APPROVED** [review]

File: test/starknet/shared/short.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
+export const SHORT_STR_SIZE = 31;
```
Comment (bonustrack) [inline test/starknet/shared/short.ts]: Could you use these functions from sx.js, I've moved the code there, you can do this like this: https://github.com/snapshot-labs/sx-api/blob/master/src/writer.ts#L1

File: test/starknet/vanilla_space.ts [hunk 5]
```diff
@@ -48,31 +49,53 @@ async function setup() {
   };
 }
 
+function setupProposeCalldata(
```
Comment (bonustrack) [inline test/starknet/vanilla_space.ts]: This function is useless, it only take argument to return it in slightly different way, you could do that directly on line 84

Comment (pscott) [reply test/starknet/vanilla_space.ts]: [Add sx-api dep; Remove setupcalldata function](https://github.com/snapshot-labs/sx-core/pull/54/commits/fecb1fc3d7d126c81e82b9063706002c758b0230)

File: test/starknet/shared/short.ts (hunk 4 above)
Comment (pscott) [reply test/starknet/shared/short.ts]: [Add sx-api dep; Remove setupcalldata function](https://github.com/snapshot-labs/sx-core/pull/54/commits/fecb1fc3d7d126c81e82b9063706002c758b0230)

**Comment by pscott** [comment]
can merge once @bonustrack approves :D

**Review by bonustrack — APPROVED** [review]

File: package.json [hunk 6]
```diff
@@ -25,6 +25,7 @@
     "@gnosis.pm/safe-contracts": "^1.3.0",
     "@gnosis.pm/zodiac": "^1.0.3",
     "starknet": "^2.5.1",
+    "sx-api": "snapshot-labs/sx-api",
```
Comment (bonustrack) [inline package.json:28]: C quoi sa

Comment (pscott) [reply package.json:28]: @snapshot-labs/sx

**Pull request merged**